Fix part of #632: Move PromotedStoryListAdapter to BindableAdater

masclot opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

Fixes part of #632: Reimplementation of PromotedStoryListAdapter as BindableAdater
Espresso tests are failing, but they were already failing. The errors with the new adapter are the same. The same tests pass when using gradle.

There was an error in MarginBindingAdapter which I'm also fixing in this cl: MarginBindingAdapter was replacing start/end margin changes for left/right margin changes. However, this replacement relied on getting the layout direction from the View being changed. When the direction is inherited and the View is not attached to a parent (recursively until top level View), the direction defaults to LTR, no matter the Locale/language. This situation happens for example in item Views within RecyclerView. The change does not replace start/end any more, but sets them through MarginLayoutParamsCompat so that it is compatible with api < 17.

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [X] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [X] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [X] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [X] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

| Current | new |
|-|-|
| ![bindable_adapter_current](https://user-images.githubusercontent.com/103062089/217546850-30d59ac3-d996-4856-9a55-43a61e533eb2.png) | ![bindable_adapter_new](https://user-images.githubusercontent.com/103062089/217546857-224a287d-b55a-4af9-8b05-e4855d78bd44.png) |

Unit tests:
![recently_played_span_test](https://user-images.githubusercontent.com/103062089/217547368-f86fa7c4-c915-4257-948c-90a8e73c3ab6.png)

Espresso Tests. The run for the new version has an extra error, but it is due to timings. In other runs it passes.
![recently_played_espresso_tests_current_vs_new](https://user-images.githubusercontent.com/103062089/217548889-21e0599e-e8b0-463d-a3f6-eeaaa93e674a.png)

@rt4914 PTAL

masclot committed:
Merge sw600 versions of profile_chooser_fragment.xml

masclot committed:
Merge branch 'oppia:develop' into merge-profile-chooser-xml

masclot committed:
Merge remote-tracking branch 'upstream/develop' into develop
Checking out the latest code in 'develop'

masclot committed:
Make RecentlyPlayed fragment use BindableAdapter instead of PromotedStoryListAdapter

masclot committed:
Make RecentlyPlayed fragment use BindableAdapter instead of PromotedStoryListAdapter

masclot committed:
Make RecentlyPlayed fragment use BindableAdapter instead of PromotedStoryListAdapter

masclot committed:
Fix formatting

masclot committed:
Fix formatting

masclot committed:
Fix formatting

masclot committed:
Fix formatting

masclot committed:
Merge branch 'develop' into move-promotedStoryListAdapter-to-bindableAdater

masclot commented:
@rt4914 PTAL

masclot committed:
Fix documentation issues

masclot committed:
Merge remote-tracking branch 'origin/move-promotedStoryListAdapter-to-bindableAdater' into move-promotedStoryListAdapter-to-bindableAdater

masclot committed:
Add RecentlyPlayedViewModel to the list of view models with resources in bazel

BenHenning commented on app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt:
```diff
@@ -282,6 +137,34 @@ class RecentlyPlayedFragmentPresenter @Inject constructor(
+        is SectionTitleViewModel -> ViewType.VIEW_TYPE_TITLE
+        else -> throw IllegalArgumentException("Encountered unexpected view model: $viewModel")
+      }
+    }
+      .registerViewDataBinder(
+        viewType = ViewType.VIEW_TYPE_TITLE,
+        inflateDataBinding = SectionTitleBinding::inflate,
+        setViewModel = SectionTitleBinding::setViewModel,
+        transformViewModel = { it as SectionTitleViewModel }
+      )
+      .registerViewDataBinder(
+        viewType = ViewType.VIEW_TYPE_PROMOTED_STORY,
+        inflateDataBinding = RecentlyPlayedStoryCardBinding::inflate,
+        setViewModel = RecentlyPlayedStoryCardBinding::setViewModel,
+        transformViewModel = { it as PromotedStoryViewModel }
+      )
+      .build()
+  }
```
```suggestion
    }.registerViewDataBinder(
      viewType = ViewType.VIEW_TYPE_TITLE,
      inflateDataBinding = SectionTitleBinding::inflate,
      setViewModel = SectionTitleBinding::setViewModel,
      transformViewModel = { it as SectionTitleViewModel }
    ).registerViewDataBinder(
      viewType = ViewType.VIEW_TYPE_PROMOTED_STORY,
      inflateDataBinding = RecentlyPlayedStoryCardBinding::inflate,
      setViewModel = RecentlyPlayedStoryCardBinding::setViewModel,
      transformViewModel = { it as PromotedStoryViewModel }
    ).build()
  }
```

Nit: bit cleaner wrapping is possible.

masclot replied:
Done

BenHenning commented on app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt:
```diff
@@ -39,189 +35,48 @@ class RecentlyPlayedFragmentPresenter @Inject constructor(
-            /* number of spaces this item should occupy = */ 1
-          }
+        return if (adapter.getItemViewType(position) == ViewType.VIEW_TYPE_TITLE.ordinal) {
+          /* number of spaces this item should occupy = */ spanCount
```
This isn't actually an argument comment. While it's an existing issue, I think we should fix this. Given that it doesn't really add much context, I suggest just removing this & the next copy of the comment.

masclot replied:
Done

BenHenning replied:
Thanks!

BenHenning commented on app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt:
```diff
@@ -39,189 +35,48 @@ class RecentlyPlayedFragmentPresenter @Inject constructor(
-    }
 
     val spanCount = activity.resources.getInteger(R.integer.recently_played_span_count)
-    promotedStoryListAdapter.setSpanCount(spanCount)
```
Just checking: was this not needed before with the SpanSizeLookup being provided?

masclot replied:
Not needed: the method just sets a private field that is not used internally. https://github.com/oppia/oppia-android/blob/3059cc7870c521e6b52b5740e6b1982dd9fe0a2d/app/src/main/java/org/oppia/android/app/home/recentlyplayed/PromotedStoryListAdapter.kt#L23

BenHenning replied:
Nice catch!

BenHenning commented on app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedViewModel.kt:
```diff
@@ -0,0 +1,162 @@
+  private var internalProfileId: Int = -1
+
+  /**
+   * Warning: it is required to call [setInternalProfileId] before accessing this property.
```
The KDoc should explain what this is first, before any such warnings.

masclot replied:
Done

BenHenning commented on app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedViewModel.kt:
```diff
@@ -0,0 +1,162 @@
+  private val translationController: TranslationController,
+) {
+
+  private var internalProfileId: Int = -1
```
It may be cleaner to pass in a ProfileId, instead, and mark it as a lateinit var to ensure that it actually gets initialized.

masclot replied:
Done

BenHenning commented on app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedViewModel.kt:
```diff
@@ -0,0 +1,162 @@
+  /**
+   * Warning: it is required to call [setInternalProfileId] before accessing this property.
+   */
+  val recentlyPlayedLiveData: LiveData<List<RecentlyPlayedItemViewModel>> by lazy {
```
Perhaps instead: ``recentlyPlayedItems``?

masclot replied:
Done

BenHenning commented on app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt:
```diff
@@ -39,189 +35,48 @@ class RecentlyPlayedFragmentPresenter @Inject constructor(
-          else -> {
-            /* number of spaces this item should occupy = */ 1
-          }
+        return if (adapter.getItemViewType(position) == ViewType.VIEW_TYPE_TITLE.ordinal) {
```
Does this actually correctly replicate all of the cases covered in the code removed above? Are there tests to cover those cases, too?

This seems like a significant simplification (which is great), but I'm not sure that it actually handles the different cases above.

masclot replied:
I think both are equivalent in the end result. I think the new version is better (apart from less code) because it does not need to change the layout manager, which causes [very] short term discrepancy between adapter and layout manager.
There is an existing test class that covers the span count: https://github.com/oppia/oppia-android/blob/087f2fe85aa01655ad03bbebbc64ab5edd13089f/app/src/test/java/org/oppia/android/app/testing/RecentlyPlayedSpanTest.kt#L99

BenHenning replied:
Thanks for checking--this is excellent. Thanks @masclot!

BenHenning reviewed (changes requested):
Thanks @masclot! Took a first pass--PTAL. Also, I think there are CI checks failing that should be fixed.

BenHenning commented:
Also, please make sure to include the 'Fixes part of #632' within the PR description.

masclot committed:
Address several review comments

masclot committed:
In RecentlyPlayedViewModel.kt make profileId not changeable by using a factory.
Turn some dependencies from AppCompatActivity to Context, so that unit tests do not require bringing up an activity/fragment.
Add Robolectric unit tests for RecentlyPlayedViewModel.kt

masclot committed:
Fix indentation issues

masclot committed:
Fix import order

masclot commented:
I added unit tests
@BenHenning PTAL

masclot committed:
Add KDoc

BenHenning commented on app/src/main/java/org/oppia/android/app/home/recentlyplayed/PromotedStoryViewModel.kt:
```diff
@@ -12,7 +12,7 @@ import org.oppia.android.domain.translation.TranslationController
 
 /** [ViewModel] for displaying a promoted story. */
 class PromotedStoryViewModel(
-  private val activity: AppCompatActivity,
```
Why is this change needed?

BenHenning replied:
Looks like this was removed in the latest changes.

BenHenning commented on app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedViewModel.kt:
```diff
@@ -18,46 +16,62 @@ import org.oppia.android.util.parser.html.StoryHtmlParserEntityType
+  /**
+   * Factory of RecentlyPlayedViewModel.
+   *
+   * Dagger's AssistedInject would do all the wiring automatically, unfortunately it is not imported
+   * in oppia-android.
```
This seems more like a PR comment, not something that we'd want to keep in code long-term.

Also, I haven't used Dagger's assisted injection before. I'm not opposed to integrating it provided:
- We can make it work with both Gradle and Bazel.
- It will interoperate correctly with Dagger Hilt, since integrating with Hilt is a long-term goal.

masclot replied:
I'll remove the PR part of the comment.
Re. your points, I have used Dagger Hilt with AssistedInject before with no problems. That said, I have no experience with Gradle and Bazel, so I cannot answer that part.

BenHenning commented on app/src/main/java/org/oppia/android/app/translation/AppLanguageResourceHandler.kt:
```diff
@@ -26,7 +26,7 @@ import javax.inject.Inject
  */
 class AppLanguageResourceHandler @Inject constructor(
-  private val activity: AppCompatActivity,
+  private val activity: Context,
```
Ditto  here. This seems incorrect--we cannot guarantee an activity is injected here without explicitly requesting that type (and this **must** be an activity for language switching to function correctly).

masclot replied:
The only reason was to be able to create robolectric tests without the burden of dealing with activities. If the test is not needed, then there is no point. I'm curious though: why does language switching does not work without activity? Can you point me to some code?

BenHenning replied:
It has to do with how the Dagger graph is set up, actually. Injecting a context makes this handler exist 1-per-app rather than 1-per-activity, so we end up in a situation where we're using resources only bound to the application. This is an issue because language switching triggers an activity recreation via:

https://github.com/oppia/oppia-android/blob/c7349c3c24f99b16c9f5115fd410fbe9c827e1d0/app/src/main/java/org/oppia/android/app/translation/AppLanguageWatcherMixin.kt#L90

And we change the activity's resources configuration dynamically based on the current language during activity creation:

https://github.com/oppia/oppia-android/blob/c7349c3c24f99b16c9f5115fd410fbe9c827e1d0/app/src/main/java/org/oppia/android/app/activity/InjectableAppCompatActivity.kt#L76

Theoretically, this can lead to using a Resources object that doesn't have the correct corresponding language. That's definitely wrong from an API perspective, but from a practical functionality perspective I'm less sure if it'd actually break in *this* specific case. Android may rely on other signals (like the default Locale object) for determining string retrieval, or perhaps it always uses the top-most activity of the process's task stack when retrieving strings. Either way, we want to keep things per-activity to increase predictability.

masclot replied:
Thanks for the explanation!

BenHenning commented on app/src/test/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedViewModelTest.kt:
```diff
@@ -0,0 +1,245 @@
+  application = RecentlyPlayedViewModelTest.TestApplication::class,
+  manifest = Config.NONE
+)
+class RecentlyPlayedViewModelTest {
```
Thanks for adding this test suite!

We generally don't test view models, and I'm not sure we want to establish that pattern now, either. The main reason for this is that I think view models will largely go away (or become trivial data objects) in the future as we move toward more code generation to reduce the fairly significant amount of boilerplate needed today for writing UIs.

We can generally test view models thoroughly through their corresponding fragments/activities which is also generally more useful since we're verifying the integration of the UI component itself (and lifecycles play a pretty critical role in view models behaving correctly).

Might it be possible to repurpose these tests & move them to RecentlyPlayedFragmentTest?

masclot replied:
I added the test because one of the PR checks was complaining. It said something like "new file without test file" or something similar.

masclot replied:
In fact, RecentlyPlayedFragmentTest already does equivalent tests

BenHenning replied:
Ah, we just exempt view models (longer term we really should remove the check for view models, but this is one of the things that's in a slightly temporary state for our static checks.

BenHenning reviewed (changes requested):
Thanks @masclot! I really like the cleanup work being done here. I have some thoughts on the new test suite, and a few questions about various other changes. PTAL.

BenHenning commented:
Also, @masclot as FYI we generally rely on the original creator of a comment thread to resolve it (as part of verifying that the thing was addressed). I've unresolved past comment threads that you marked as resolved, so nothing to change here (but please keep this in mind for future comment threads :)).

seanlip commented:
HI @masclot -- just to check, are you still working on this? Thanks!

masclot commented:
Yes. I was out last week.

seanlip commented:
Ah, OK. Thanks!

masclot committed:
Remove the unit tests added in previous commit, plus undo changes required for that test to run.

masclot committed:
Fix MarginBindingAdapters to work well when the target View is not yet attached, and thus information about the layout direction is not available.

masclot commented:
I fixed the MarginBindingAdapter class as explained in the last meeting. I'm attaching some screenshots of the home, select_profile and recently_played screens in RTL model.

| Current | New |
| - | - |
| Recently played ||
| ![recently_played_current](https://user-images.githubusercontent.com/103062089/222734433-c5625254-0276-4982-a390-915e4c922501.png) | ![recently_played_new](https://user-images.githubusercontent.com/103062089/222734438-ecd625ba-5e7b-4a35-8cb2-64727fc1ba8c.png) |
| Home ||
| ![home_current](https://user-images.githubusercontent.com/103062089/222734543-8cb01d4e-7fc3-4bc2-8e9d-875901c15ebb.png) | ![home_new](https://user-images.githubusercontent.com/103062089/222734548-56a4a18d-381b-423c-9ccc-cd4524ab5504.png) |
| Select profile ||
| ![select_profile_current](https://user-images.githubusercontent.com/103062089/222734624-05547088-9c10-4fd7-95b7-3aff8ac84041.png) | ![select_profile_new](https://user-images.githubusercontent.com/103062089/222734626-036ee8d5-76ca-4c6f-a744-51716b5568f1.png) |

MarginBindingAdapters tests:
![MarginBindingAdapters tests](https://user-images.githubusercontent.com/103062089/222734903-82d8f1b0-3c16-4353-9d3f-b789f5ad37b3.png)

masclot commented:
@BenHenning PTAL

adhiamboperes commented on app/src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java:
```diff
@@ -14,12 +14,7 @@ public final class MarginBindingAdapters {
-      } else {
-        setLayoutDirectionalMargins(view, (int) marginStart, (int) marginEnd);
-      }
+      MarginLayoutParamsCompat.setMarginStart(params, (int) marginStart);
```
Could you add an explanation in your PR description for why we made the change to `MarginLayoutParamsCompat`

masclot replied:
Done

adhiamboperes commented on app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt:
```diff
@@ -39,199 +35,57 @@ class RecentlyPlayedFragmentPresenter @Inject constructor(
   private val routeToExplorationListener = activity as RouteToExplorationListener
-  private var internalProfileId: Int = -1
+
+  private lateinit var internalProfileId: ProfileId
```
`internalProfileId: Int = -1` is usually the default value used when profileId is not initialized yet or not found, and I am concerned that we might be introducing a nullability issue by  lateinit-ing internalProfileId here.

masclot replied:
Changed back to -1

adhiamboperes replied:
Took another pass. This will actually work, and there's no need to change it.

masclot replied:
Changed back to lateinit

adhiamboperes reviewed (commented):
You can format the images in the PR [current,new] next to each other by using tables. Here is [how to add table syntax in md](https://www.markdownguide.org/extended-syntax/). The outcome would be something like [this](https://github.com/oppia/oppia-android/issues/3345#issuecomment-1317551932).

BenHenning commented on app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedViewModel.kt:
```diff
@@ -0,0 +1,171 @@
+import javax.inject.Inject
+
+/** View model for [RecentlyPlayedFragment]. */
+class RecentlyPlayedViewModel constructor(
```
We don't need to explicitly declare the constructor unless changing its visibility. That being said, if we can make this private that'd be preferred since the pattern here is to use the view model's factory to create instances of it.

masclot replied:
Set to private

BenHenning replied:
Resolved in the latest changes.

BenHenning commented on app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedViewModel.kt:
```diff
@@ -0,0 +1,171 @@
+    }
+  }
+
+  private fun processPromotedStoryList(promotedActivityList: PromotedActivityList):
```
You can break at the ``(`` instead to avoid the extra indentation being needed here (in the same way as ``addRecentlyPlayedStoryListSection`` below).

masclot replied:
Done

BenHenning replied:
Resolved in the latest changes.

BenHenning reviewed (changes requested):
Thanks @masclot! Just had a few follow-up comments.

masclot commented:
> You can format the images in the PR [current,new] next to each other by using tables. Here is [how to add table syntax in md](https://www.markdownguide.org/extended-syntax/). The outcome would be something like [this](https://github.com/oppia/oppia-android/issues/3345#issuecomment-1317551932).

Thanks. I updated the comments as indicated.

masclot committed:
Address PR comments

masclot committed:
Address PR comments

adhiamboperes commented on app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt:
```diff
@@ -39,199 +35,57 @@ class RecentlyPlayedFragmentPresenter @Inject constructor(
-      activity,
-      promotedStory,
-      entityType,
+    this.internalProfileId = ProfileId.newBuilder().setInternalId(profileId).build()
```
@masclot, It looks like you're building another instance of ProfileId here, but I think you should assign the variable created earlier on line 46.

masclot replied:
I'm not sure I understand. I'm creating a ProfileId in order to assign to the variable created on line 46. The variable on line 46 holds a fully built proto, not a proto-builder.

Changed back to lateinit

adhiamboperes replied:
@masclot, tagging you here since this is the comment I was referring to on chat.

BenHenning replied:
Nit: we also ought to call this ``profileId`` since it's not an "internal" profile ID anymore. Besides that, it should be fine for this variable to be a ``lateinit var``--no need to initialize it as @adhiamboperes suggests here (in case the most recent change was due to an old comment of mine--I'm not sure if that's the case here).

masclot replied:
Done

adhiamboperes reviewed (changes requested):
I've left one code comment blocking approval.

Could you also PTAL @masclot at the failing script check and resolve it? Let me know if you need any help with that.

masclot commented:
> I've left one code comment blocking approval.
> 
> Could you also PTAL @masclot at the failing script check and resolve it? Let me know if you need any help with that.

The failing check is due to a new class file without a matching test file. Since the new class is a ViewModel, @BenHenning  told me I didn't need to add a test file.

masclot commented:
PTAL @adhiamboperes

adhiamboperes commented:
> > I've left one code comment blocking approval.
> > Could you also PTAL @masclot at the failing script check and resolve it? Let me know if you need any help with that.
> 
> The failing check is due to a new class file without a matching test file. Since the new class is a ViewModel, @BenHenning told me I didn't need to add a test file.

For files we are exempting from tests, we add them to `test_file_exemptions.textproto`. Note that the same lexicographical ordering rule of imports applies here too.

BenHenning commented on app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedViewModel.kt:
```diff
@@ -0,0 +1,172 @@
+  private val resourceHandler: AppLanguageResourceHandler,
+  private val translationController: TranslationController,
+  private val promotedStoryClickListener: PromotedStoryClickListener,
+  private val internalProfileId: ProfileId,
```
Nit: ``profileId``.

masclot replied:
Done both

BenHenning reviewed (changes requested):
Thanks @masclot! The PR mostly LGTM, just had a few follow-ups.

Also, please make sure to reply to all comments before sending the PR back to review (otherwise it can be difficult as a reviewer to know what you've intended to address).

BenHenning commented:
(Taking off Rajat since he doesn't need to review this PR).

masclot commented:
> > > I've left one code comment blocking approval.
> > > Could you also PTAL @masclot at the failing script check and resolve it? Let me know if you need any help with that.
> > 
> > 
> > The failing check is due to a new class file without a matching test file. Since the new class is a ViewModel, @BenHenning told me I didn't need to add a test file.
> 
> For files we are exempting from tests, we add them to `test_file_exemptions.textproto`. Note that the same lexicographical ordering rule of imports applies here too.

Done

masclot committed:
Added Test exemption for RecentlyPlayedViewModel.
Renamed field internalProfileId to profileId; also, make it lateinit.

adhiamboperes commented:
> > > > I've left one code comment blocking approval.
> > > > Could you also PTAL @masclot at the failing script check and resolve it? Let me know if you need any help with that.
> > > 
> > > 
> > > The failing check is due to a new class file without a matching test file. Since the new class is a ViewModel, @BenHenning told me I didn't need to add a test file.
> > 
> > 
> > For files we are exempting from tests, we add them to `test_file_exemptions.textproto`. Note that the same lexicographical ordering rule of imports applies here too.
> 
> Done

Thanks @masclot. The PR LGTM. Can you respond to all the individual comments above that you have addressed, so we can mark them as resolved?

adhiamboperes reviewed (approved):
Aside from unresolved comments, this generally looks good.

masclot commented:
> Aside from unresolved comments, this generally looks good.

I cannot find any unresolved comment I haven't replied to. Is there a View in github that summarizes that lists the unresolved comments? Or maybe you can somehow highlight those comments for me/link to them?

masclot commented:
When I click on "Conversations" it shows three unresolved comments and I replied to all of them.

seanlip commented:
@masclot [This one](https://github.com/oppia/oppia-android/pull/4874#discussion_r1144089394) and [this one](https://github.com/oppia/oppia-android/pull/4874#discussion_r1144089796)? (I found these by scrolling up on this page.)

masclot commented:
Thanks Sean. The first one does not have a "Reply" box. The second one I already replied to with "Done both", trying to include the comment I cannot reply to. This comment has a "Pending" label; I don't know what that means.

seanlip commented:
Ah, for the first one, scroll up the page further and click "Load more". Here's the start of that thread: https://github.com/oppia/oppia-android/pull/4874#discussion_r1143479652 -- you can reply there.

If the comment box says Pending, that means that you have draft comments that haven't been submitted. Go to the "Files Changed" tab, and submit them using the green button in the top right.

masclot commented:
Thanks! I thought that big green button was for reviewers, not for replies to review-comments. I also don't understand why some of my comments went through without clicking that button. 
Anyway, there are no more pending labels and hopefully no more unreplied comments.

seanlip commented:
Ah ok, great, thanks @masclot! Not sure about why some of the comments went through, that seems odd.

Also just wondering, do you have any improvements to suggest to the docs [here](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#step-3-address-review-comments-until-all-reviewers-give-lgtm) that would have made the process clearer? If so, would you be up for making a PR to update them? (Now that the wiki docs are stored in the main repo, that should hopefully be straightforward.)

Thanks!

adhiamboperes commented:
> Ah ok, great, thanks @masclot! Not sure about why some of the comments went through, that seems odd.
> 
> Also just wondering, do you have any improvements to suggest to the docs [here](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#step-3-address-review-comments-until-all-reviewers-give-lgtm) that would have made the process clearer? If so, would you be up for making a PR to update them? (Now that the wiki docs are stored in the main repo, that should hopefully be straightforward.)
> 
> Thanks!

cc @seanlip. @masclot, I believe what happened here was: in some cases you replied from the conversation view, which adds a single comment, and in a later instance, you chose to 
*start a review*, from the *Files changed* tab.

**This**
<img width="1360" alt="Screenshot 2023-03-29 at 14 03 47" src="https://user-images.githubusercontent.com/59600948/228514388-571b041c-24e4-4fb0-b977-a2d10b934972.png">

**vs**

<img width="1332" alt="Screenshot 2023-03-29 at 14 03 08" src="https://user-images.githubusercontent.com/59600948/228514445-461dd428-bd98-42ce-a950-a325cfa1f769.png">

BenHenning reviewed (approved):
Thanks @masclot! This LGTM!

BenHenning commented:
Syncing with develop & enabling auto-merge.

BenHenning committed:
Merge branch 'develop' into move-promotedStoryListAdapter-to-bindableAdater

masclot commented:
> cc @seanlip. @masclot, I believe what happened here was: in some cases you replied from the conversation view, which adds a single comment, and in a later instance, you chose to
> _start a review_, from the _Files changed_ tab.

Thant makes sense, thanks

masclot commented:
> Also just wondering, do you have any improvements to suggest to the docs [here](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#step-3-address-review-comments-until-all-reviewers-give-lgtm) that would have made the process clearer? If so, would you be up for making a PR to update them? (Now that the wiki docs are stored in the main repo, that should hopefully be straightforward.)

I'll try to compile a list of things I had trouble with. The documentation as a whole is big; I'm sure it has answers to at least some of my troubles, but I just could not remember the solution when I faced them.

seanlip commented:
@masclot Sounds good, thanks a lot -- appreciate it! 

/cc @MohitGupta121 too, who is doing some work to update the wiki to make it easier to follow.

MohitGupta121 commented:
Thanks @seanlip, Yes @masclot glad to help and collaborate with you in wiki contribution.
